ingestion/functions/parsing/germany/germany.py

```diff
@@ -0,0 +1,105 @@
+        os.path.join(
+            os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "common/python"
+        )
+    )
+import parsing_lib
+
+_DATE_INDEX = 8
+_AGE_INDEX = 4
+_GENDER_INDEX = 5
+_CASECOUNT = 6
+_DEATHCOUNT = 7
+_RECOVERYCOUNT = 15
+_ADMIN1 = 2
+_ADMIN2 = 3
+
+
+def convert_date(raw_date):
+    """
+    Convert raw date field into a value interpretable by the dataserver.
+    """
+    date = datetime.strptime(raw_date, "%Y/%m/%d %H:%M:%S")
+    return date.strftime("%m/%d/%YZ")
+
+
+def convert_gender(raw_gender):
+    if raw_gender == "M":
+        return "Male"
+    if raw_gender == "W":
+        return "Female"
+
+
+def convert_demographics(gender: str, age: str):
+    demo = {}
+    if gender:
+        demo["gender"] = convert_gender(gender)
+    if age:
+        age_range = age.split("-")
+        if len(age_range) > 1:
+            demo["ageRange"] = {
+                "start": float(age_range[0][1:3]),
+                "end": float(age_range[1][1:3]),
+            }
+        else:
+            demo["ageRange"] = {
+                "start": float(age_range[0][1:3]),
+                "end": float(age_range[0][1:3]),
+            }
+    return demo
+
+
+def parse_cases(raw_data_file: str, source_id: str, source_url: str):
+    """
+    Parses G.h-format case data from raw API data.
+    """
+    with open(raw_data_file, "r") as f:
+        reader = csv.reader(f)
+        next(reader)  # Skip the header.
+        cases = []
+        for row in reader:
+            num_confirmed_cases = int(row[_CASECOUNT])
+            if num_confirmed_cases < 1:
+                continue
+            try:
+                case = {
+                    "caseReference": {"sourceId": source_id, "sourceUrl": source_url},
+                    "location": {
+                        "query": ", ".join(("Germany", row[_ADMIN1], row[_ADMIN2])),
+                        "limitToResolution": "Admin2",
+                    },
+                    "events": [
+                        {
+                            "name": "confirmed",
+                            "dateRange": {
+                                "start": convert_date(row[_DATE_INDEX]),
+                                "end": convert_date(row[_DATE_INDEX]),
+                            },
+                        },
+                    ],
+                    "demographics": convert_demographics(
+                        row[_GENDER_INDEX], row[_AGE_INDEX]
+                    ),
+                }
+                # Death and recovery counts are always equal to case counts for the row.
+                if int(row[_DEATHCOUNT]) > 0:
+                    case["events"].append({"name": "outcome", "value": "Death"})
+                elif int(row[_RECOVERYCOUNT]) > 0:
+                    case["events"].append({"name": "outcome", "value": "Recovered"})
+                cases.extend([case] * num_confirmed_cases)
+            except ValueError as ve:
+                print(ve)
```

Comment:
Downside to catching these is that we won't know if we're missing substantial (but human-review-missable) amounts of data. If the quality of this data is such that we need to handle occasional errors like this, that's OK, but we should probably drop a TODO to figure out a monitoring solution for this.

Reply:
Sounds good! I just tested this on the latest dataset, and it actually doesn't throw any errors at all - just kept this in there following the example of Zurich since it seemed like a good idea just in case. 👍

Reply:
zurich was my first parser, for new ones I actually end up raising ValueErrors more ofter, I think we can raise as well here.